File: snippets/go/services/cancellation-example.go
package main

import (
	restate "github.com/restatedev/sdk-go"
)

// Type definitions
type Order struct {
	UserID string
	Amount float64
	ID     string
}

type Payment struct {
	ID     string
	Amount float64
}

// Service stubs
var notificationService = struct {
	Notify func(userID, message string) error
}{
	Notify: func(userID, message string) error {
		// Simulate notification
		return nil
	},
}

// Function stubs
func processPayment(id string, order Order) (Payment, error) {
	// Simulate payment processing
	return Payment{
		ID:     "payment_" + order.ID,
		Amount: order.Amount,
	}, nil
}

func refundPayment(id string) error {
	// Simulate refund processing
	return nil
}

type OrderService struct {
}

// <start_here>
func (OrderService) processOrder(ctx restate.ObjectContext, order Order) error {
	// If cancellation happened right before this line, this still executes
	restate.Set(ctx, "status", "processing")

	// If cancellation happened right before this line, this still executes
	paymentId := restate.Rand(ctx).UUID().String()

	// If cancelled right before this line, Run won't execute
	// If cancelled during run block execution,
	// then a terminal error gets returned here once execution finishes
	_, err := restate.Run(ctx, func(ctx restate.RunContext) (Payment, error) {
		return processPayment(paymentId, order)
	})
	if err != nil {
		if restate.IsTerminalError(err) {
			// Cancellation detected - run compensation
			_, _ = restate.Run(ctx, func(ctx restate.RunContext) (any, error) {
				return nil, refundPayment(paymentId)
			})
		}
		return err
	}

	// If cancellation happened right before this line, this still executes
	restate.ServiceSend(ctx, "NotificationService", "Notify").
		Send("Payment processed")

	return nil
}

// <end_here>

File: snippets/python/src/services/cancellation_example.py
import restate
from restate.exceptions import TerminalError
def process_payment(id, order):
    pass


def refund_payment(id):
    pass

order_service = restate.VirtualObject("order-service")

class Order:
    pass

@order_service.handler()
async def notify(ctx: restate.ObjectContext, message: str):
    pass

# <start_here>
@order_service.handler()
async def process_order(ctx: restate.ObjectContext, order: Order):
    # If cancellation happened before this line, this still executes
    ctx.set("status", "processing")

    # If cancellation happened before this line, this still executes
    payment_id = str(ctx.uuid())
    try:
        # If cancelled before this await, ctx.run won't execute
        # If cancelled during run block execution,
        # then a terminal error gets raised here once execution finishes
        await ctx.run_typed("process-payment", process_payment, id=payment_id, order=order)

        # If cancellation happened before this line, this still executes
        ctx.service_send(notify, "Payment processed")

    except TerminalError as e:
        # Cancellation detected - run compensation
        await ctx.run_typed("refund-payment", refund_payment, id=payment_id)
        raise e # Re-throw to propagate cancellation
# <end_here>
